src/QuartoGameState.java

```diff
@@ -160,45 +164,100 @@ public void evaluate(QuartoPiece limboPiece) {
         }
         else {
             value = 0;
-            // FIXME/OPTIMIZE: this is just hacky and awful
-            for(int[] square : this.freeSquares) {
+            //Minor optimization, don't bother checking if we wouldn't have
+            //a board with 5 or more pieces
+            if(this.freeSquares.size() <= 21) {
+                for (int[] square : this.getTestSquares()) {
+                    QuartoBoard tmpBoard = new QuartoBoard(this.board);
+                    tmpBoard.insertPieceOnBoard(square[0], square[1], limboPiece.getPieceID());
+
+                    QuartoGameState state = getRegisteredState(tmpBoard, alpha, beta, !this.isMaxState);
+
+                    if (state.hasQuarto()) {
+                        value = state.isMaxState ? -27 : 27;
+                        break;
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Tries to find a piece that cannot be placed on the
+     * board to win.
+     *
+     * @return Proposed QuartoPiece
+     */
+    public QuartoPiece getSafePiece() {
```

Comment:
We might be able to use these methods inside of our QGTG so that we avoid bad paths.